syz-manager/snapshot.go

```diff
@@ -0,0 +1,172 @@
+	// So we start a background process and log to /dev/kmsg.
+	cmd := fmt.Sprintf("nohup %v exec snapshot 1>/dev/null 2>/dev/kmsg </dev/null &", executor)
+	if _, _, err := inst.Run(time.Hour, mgr.reporter, cmd); err != nil {
+		return err
+	}
+
+	builder := flatbuffers.NewBuilder(0)
+	var envFlags flatrpc.ExecEnv
+	for first := true; ; first = false {
+		queue.StatExecs.Add(1)
+		req := mgr.source.Next()
+		if first {
+			envFlags = req.ExecOpts.EnvFlags
+			if err := mgr.snapshotSetup(inst, builder, envFlags); err != nil {
+				return err
+			}
+		}
+		if envFlags != req.ExecOpts.EnvFlags {
+			return fmt.Errorf("request env flags has changed: 0x%x -> 0x%x",
+				envFlags, req.ExecOpts.EnvFlags)
+		}
+
+		res, output, err := mgr.snapshotRun(inst, builder, req)
+		if err != nil {
+			return err
+		}
+
+		if mgr.reporter.ContainsCrash(output) {
+			res.Status = queue.Crashed
+			rep := mgr.reporter.Parse(output)
+			buf := new(bytes.Buffer)
+			fmt.Fprintf(buf, "program:\n%s\n", req.Prog.Serialize())
+			buf.Write(rep.Output)
+			rep.Output = buf.Bytes()
+			mgr.crashes <- &Crash{Report: rep}
+		}
+
+		req.Done(res)
+	}
+}
+
+func (mgr *Manager) snapshotSetup(inst *vm.Instance, builder *flatbuffers.Builder, env flatrpc.ExecEnv) error {
+	msg := flatrpc.SnapshotHandshakeT{
+		CoverEdges:       mgr.cfg.Experimental.CoverEdges,
+		Kernel64Bit:      mgr.cfg.SysTarget.PtrSize == 8,
+		Slowdown:         int32(mgr.cfg.Timeouts.Slowdown),
+		SyscallTimeoutMs: int32(mgr.cfg.Timeouts.Syscall / time.Millisecond),
+		ProgramTimeoutMs: int32(mgr.cfg.Timeouts.Program / time.Millisecond),
+		Features:         mgr.enabledFeatures,
+		EnvFlags:         env,
+		SandboxArg:       mgr.cfg.SandboxArg,
+	}
+	builder.Reset()
+	builder.Finish(msg.Pack(builder))
+	return inst.SetupSnapshot(builder.FinishedBytes())
+}
+
+func (mgr *Manager) snapshotRun(inst *vm.Instance, builder *flatbuffers.Builder, req *queue.Request) (
+	*queue.Result, []byte, error) {
+	progData, err := req.Prog.SerializeForExec()
+	if err != nil {
+		queue.StatExecBufferTooSmall.Add(1)
+		return &queue.Result{Status: queue.ExecFailure}, nil, nil
+	}
+	msg := flatrpc.SnapshotRequestT{
+		ExecFlags: req.ExecOpts.ExecFlags,
+		NumCalls:  int32(len(req.Prog.Calls)),
+		ProgData:  progData,
+	}
+	for _, call := range req.ReturnAllSignal {
+		if call < 0 {
+			msg.AllExtraSignal = true
+		} else {
+			msg.AllCallSignal |= 1 << call
+		}
+	}
+	builder.Reset()
+	builder.Finish(msg.Pack(builder))
+
+	start := time.Now()
+	res, output, err := inst.RunSnapshot(builder.FinishedBytes())
+	if err != nil {
+		return nil, nil, err
+	}
+	elapsed := time.Since(start)
+	queue.StatExecs.Add(1)
+
+	execError := ""
+	var info *flatrpc.ProgInfo
+	if len(res) > 4 {
+		res = res[4:]
+		var raw flatrpc.ExecutorMessageRaw
+		raw.Init(res, flatbuffers.GetUOffsetT(res))
+		union := raw.UnPack()
+		if union.Msg == nil || union.Msg.Value == nil {
+			return nil, nil, errors.New("received no message")
```

Comment:
Could we land here also in the case of some executor memory corruption?

Reply:
Changed handling to treat this as no info at all (we will return queue.Result w/o program info).